napari/_vispy/overlays/scale_bar.py

```diff
@@ -65,13 +67,25 @@ def _calculate_best_length(self, desired_length: float):
         # that might have occurred (e.g. um -> cm)
         factor = current_quantity / new_quantity
 
-        # select value closest to one of our preferred values
-        index = bisect.bisect_left(PREFERRED_VALUES, new_quantity.magnitude)
+        # select value closest to one of our preferred values and also
+        # validate if quantity is dimensionless and lower than 1 to prevent
+        # the scale bar to extend beyond the canvas when zooming.
+        # If the value falls in those conditions, we use the corresponding
+        # prefered value but scaled to take into account the actual value
+        # magnitude. See https://github.com/napari/napari/issues/5914
+        magnitude_1000 = floor(log(new_quantity.magnitude, 1000))
+        scaled_magnitude = new_quantity.magnitude * 1000 ** (-magnitude_1000)
+        index = bisect.bisect_left(PREFERRED_VALUES, scaled_magnitude)
         if index > 0:
             # When we get the lowest index of the list, removing -1 will
             # return the last index.
             index -= 1
         new_value = PREFERRED_VALUES[index]
+        if new_quantity.dimensionless and new_quantity.magnitude < 1:
+            new_value = float(
+                Decimal(PREFERRED_VALUES[index])
```

Comment:
```suggestion
                Decimal(new_value)
```
?

I believe that black might be un-happy, so we might need to reformat. And I would maybe add a small comment here along:

```
# using Decimal necessary to avoid `4.999999e-6` at really small scale.
```

Suggestions, but completely unblocking, for me this is good to go as it. 

This is great !

Reply:
Thank you so much for all the feedback and help with this!